### DISCUSSION TITLE
Replicating FastAPI's Dependency Injection in pure Python

### DISCUSSION THREAD

**Discussion by navhits** [fastapi/fastapi#4192] [Questions] [answered]
The dependency injection implementation in FastAPI is clearly an outstanding design. Since Python doesn't directly allow this, there would be a way around implementing this feature. I tried to read through the code and check the call stack but don't seem to get there. If any of you would have idea around how this is done, it will be great.

Ref: https://fastapi.tiangolo.com/tutorial/dependencies/

**Comment by sureshdsk** [comment, marked as answer]
With a bit of going through fastapi code and some research, I've come up with the code below. this is merely a proof of concept and not recommended to use as it is. tested on python 3.9

```python
from typing import Callable
from functools import wraps
import inspect


class Depends:
    def __init__(self, dependency: Callable):
        self.dependency = dependency

    def __call__(self, *args, **kwargs):
        return self.dependency()


def get_default_name():
    return "Guest"


def resolve_dependency(func):
    @wraps(func)
    def wrapped(*args, **kwargs):
        sig = inspect.signature(func)
        for param in sig.parameters.values():
            if param.default is not param.empty:
                kwargs[param.name] = param.default()
        return func(*args, **kwargs)
    return wrapped


@resolve_dependency
def test_hello(name: str = Depends(get_default_name)):
    print(f"Hello {name}")


if __name__ == "__main__":
    test_hello()

```

@tiangolo might have an better idea

**Reply by navhits** [reply]
Since here we don't pass arguments we could just pass get_default_name() to the function parameter itself. I'm figuring out on a way to pass the arguments now.

**Reply by sureshdsk** [reply]
We can modify my code to support arguments just like the one in fastapi 
https://fastapi.tiangolo.com/advanced/advanced-dependencies/

**Reply by navhits** [reply]
I think this is the simplest way to achieve what Dependency injection can do. I am further looking at ways to pass the calling method arguments to the dependant module. However, that is beyond our purpose here. I marking this as the answer for the discussion for now.

**Comment by billcrook** [comment]
You might be interested in https://github.com/ets-labs/python-dependency-injector/

**Reply by navhits** [reply]
Thanks. Seen this. For some reason it doesn't look very scalable to me
